arangod/Pregel/Algos/HITSKleinberg.cpp

```diff
@@ -0,0 +1,333 @@
+    switch (state) {
+      case State::sendInitialHubs: {
+        state = State::updateAuth;
+      } break;
+      case State::updateAuth:
+      case State::updateAuthNormalizeHub: {
+        state = State::updateHubNormalizeAuth;
+      } break;
+      case State::updateHubNormalizeAuth: {
+        ++currentIteration;
+        if (currentIteration == numIterations) {
+          state = State::finallyNormalizeHubs;
+        } else {
+          state = State::updateAuthNormalizeHub;
+        }
+      } break;
+      case State::finallyNormalizeHubs:
+      case State::finallyNormalizeAuths: {
+        aggregate(isLastIterationAggregator, true);
+      } break;
+    }
+  }
+};
+
+struct HITSKleinbergComputation
+    : public VertexComputation<VertexType, int8_t, SenderMessage<double>> {
+  HITSKleinbergComputation() = default;
+
+  // sends auth to all in-neighbors. note that all vertices send messages
+  // to all out-neighbors in all iterations (there are no inactive vertices),
+  // so the set of in-neighbors can be determined by iterating over received
+  // messages
+  void sendAuthToInNeighbors(
+      MessageIterator<SenderMessage<double>> const& receivedMessages,
+      double auth) {
+    SenderMessage<double> authData(this->pregelId(), auth);
+    for (SenderMessage<double> const* message : receivedMessages) {
+      sendMessage(message->senderId, authData);
+    }
+  }
+
+  void sendHubToOutNeighbors(double hub) {
+    SenderMessage<double> hubData(this->pregelId(), hub);
+    sendMessageToAllNeighbours(hubData);
+  }
+
+  void compute(
+      MessageIterator<SenderMessage<double>> const& messages) override {
+    auto const* ctx =
+        dynamic_cast<HITSKleinbergWorkerContext const*>(context());
+    switch (ctx->state) {
+      case State::sendInitialHubs: {
+        // these are not normalized, but according to the description of the
+        // algorithm in the paper, the 1.0's are used in place of normalized
+        // values
+        mutableVertexData()->normalizedAuth = 1.0;
+        mutableVertexData()->normalizedHub = 1.0;
+        reportFakeDifference();
+
+        sendHubToOutNeighbors(1.0);
+      } break;
+
+      case State::updateAuth: {
+        // We enter this state when all authorities and all hubs are 1.0
+        updateStoreAndSendAuth(messages, 1.0);
+        reportFakeDifference();
+        // authorities are one iteration before hubs
+        // authorities are not normalized, hubs are 1.0
+
+        // Next state: updateHubNormalizeAuth
+      } break;
+
+      case State::updateHubNormalizeAuth: {
+        // authorities are updated one iteration more than hubs,
+        // authorities are not normalized, hubs are normalized
+
+        // update local hub
+        double nonNormalizedHub = 0.0;
+        for (SenderMessage<double> const* message : messages) {
+          TRI_ASSERT(message != nullptr);
+          nonNormalizedHub += message->value;  // auth from our out-neighbors
+        }
```

Comment:
Note for the future: let's look at fold expressions to write this more concisely